app/src/visionglass/java/com/igalia/wolvic/PlatformActivity.java

```diff
@@ -113,54 +116,94 @@ private void runPhoneUICallback(PhoneUIButtonsCallback callback) {
+        mSensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
+
+        // Alternatively: android.hardware.usb.action.USB_DEVICE_ATTACHED, USB_DEVICE_DETACHED.
+        IntentFilter usbPermissionFilter = new IntentFilter();
+        usbPermissionFilter.addAction(HUAWEI_USB_PERMISSION);
+        registerReceiver(mUsbPermissionReceiver, usbPermissionFilter);
+
         VisionGlass.getInstance().init(getApplication());
+        VisionGlass.getInstance().setOnConnectionListener(this);
+        initVisionGlass();
+    }
 
-        mDisplayManager = (DisplayManager) getSystemService(Context.DISPLAY_SERVICE);
-        boolean wasImuStarted = false;
-        boolean isAskingForPermission = false;
-        do {
-            if (VisionGlass.getInstance().isConnected()) {
-                if (VisionGlass.getInstance().hasUsbPermission()) {
-                    Log.d(LOGTAG, "Device has USB permission -> registering callback for startImu");
-                    wasImuStarted = true;
-                    VisionGlass.getInstance().startImu((w, x, y, z) -> queueRunnable(() -> setHead(x, y, z, w)));
-                } else {
-                    if (isAskingForPermission) {
-                        try {
-                            Thread.sleep(100);
-                        } catch (InterruptedException e) {
-                            e.printStackTrace();
-                        }
-                    } else {
-                        Log.w(LOGTAG, "Device does not have USB permission -> asking");
-                        VisionGlass.getInstance().requestUsbPermission();
-                        isAskingForPermission = true;
-                    }
-                }
-            } else {
-                // TODO: show a dialog asking the user to put on the glasses
-                Log.e(LOGTAG, "Device not connected" + (VisionGlass.getInstance().hasUsbPermission() ? " and does NOT have USB permissions" : ""));
+    @Override
+    public void onConnectionChange(boolean b) {
+        if (VisionGlass.getInstance().isConnected()) {
+            Log.d(LOGTAG, "onConnectionChange: Device connected");
+            initVisionGlass();
+        } else {
+            Log.d(LOGTAG, "onConnectionChange: Device disconnected");
+            if (mWasImuStarted) {
+                Log.d(LOGTAG, "onConnectionChange: Finish activity");
+                finish();
+            }
+        }
+    }
+
+    private void initVisionGlass() {
+        Log.d(LOGTAG, "initVisionGlass");
+
+        if (mWasImuStarted) {
+            Log.d(LOGTAG, "Duplicated call to init the Vision Glass system");
+            updateDisplays();
+            return;
+        }
+
+        if (!VisionGlass.getInstance().isConnected()) {
+            Log.d(LOGTAG, "Glasses not connected yet");
+            return;
+        }
+
+        if (VisionGlass.getInstance().hasUsbPermission()) {
+            mIsAskingForPermission = false;
```

Comment:
I think we should initialize it to false and remove this from here. Perhaps you'd like to reset it to false in the USB permission request handler.

Reply:
Fixed.